kmcd.md

```diff
@@ -87,30 +87,189 @@ End Semantics
+         <end-live> false </end-live>
+         <end-tag> TAGS => TAGS [ ILK <- PAR /Rat PIP ] </end-tag>
+         <end-art> ARTS => ARTS [ ILK <- ART ] </end-art>
+         <spot-par> PAR </spot-par>
+         <spot-ilks>
+           ...
+           ILK |-> SpotIlk(... pip: PIP)
+           ...
+         </spot-ilks>
+         <vat-ilks>
+           ...
+           ILK |-> Ilk(... Art: ART)::VatIlk
+           ...
+         </vat-ilks>
+       requires notBool ILK in_keys(TAGS)
 
-    syntax EndStep ::= "skim" Int Address
- // -------------------------------------
+    syntax EndStep ::= "skip" String Int
+ // ------------------------------------
+    rule <k> End . skip ILK ID
+          => call Vat . suck address(Vow) address(Vow) TAB
+          ~> call Vat . suck address(Vow) THIS BID
+          ~> call Vat . hope address(Flip ILK)
+	  ~> call Flip ILK . yank ID
+          ~> call Vat . grab ILK USR THIS address(Vow) LOT (TAB /Rat RATE) ... </k>
+         <this> THIS </this>
+         <end-tag>
+          ...
+          ILK |-> TAG
+          ...
+         </end-tag>
+         <end-art>
+           ...
+           ILK |-> (ART => ART +Rat (TAB /Rat RATE))
+           ...
+         </end-art>
+         <vat-ilks>
+           ...
+           ILK |-> Ilk(... rate: RATE)::VatIlk
+           ...
+         </vat-ilks>
+         <flip-ilk> ILK </flip-ilk>
+         <flip-bids>
+           ...
+           ID |-> Bid(... bid: BID, lot: LOT, usr: USR, tab: TAB)
+           ...
+         </flip-bids>
+      requires TAG =/=Rat 0
+       andBool LOT >=Rat 0
+       andBool TAB /Rat RATE >=Rat 0
 
-    syntax EndStep ::= "free" Int
- // -----------------------------
+    syntax EndStep ::= "skim" String Address
+ // ----------------------------------------
+    rule <k> End . skim ILK URN
+          => call Vat . grab ILK URN THIS address(Vow) (0 -Rat minRat(INK, ART *Rat RATE *Rat TAG)) (0 -Rat ART) ... </k>
+         <this> THIS </this>
+         <end-tag>
+          ...
+          ILK |-> TAG
+          ...
+         </end-tag>
+         <end-gap>
+          ...
+          ILK |-> (GAP => GAP +Rat ((ART *Rat RATE *Rat TAG) -Rat minRat(INK, (ART *Rat RATE *Rat TAG))))
+          ...
+         </end-gap>
+         <vat-ilks>
+           ...
+           ILK |-> Ilk(... rate: RATE)::VatIlk
+           ...
+         </vat-ilks>
+         <vat-urns>
+           ...
+           {ILK, URN} |-> Urn(... ink: INK, art: ART)
+           ...
+         </vat-urns>
+      requires TAG =/=Rat 0
+
+    syntax EndStep ::= "free" String
+ // --------------------------------
+    rule <k> End . free ILK
+          => call Vat . grab ILK MSGSENDER MSGSENDER address(Vow) (0 -Rat INK) 0 ... </k>
+         <msg-sender> MSGSENDER </msg-sender>
+         <end-live> false </end-live>
+         <vat-urns>
+           ...
+           {ILK, MSGSENDER} |-> Urn(... ink: INK, art: ART)
+           ...
+         </vat-urns>
+      requires ART ==Int 0
```

Comment:
This is actually safe because the rule will only apply if ART is of sort Int, and 0 is an Int. So it's actually slightly faster to use ==Int here instead of ==Rat, although it would not be correct if this were =/=Int.